esm-loader.mjs

```diff
@@ -66,22 +73,80 @@ export async function resolve(specifier, context, nextResolve) {
       logger.debug(
         `Registered CommonJS instrumentation for ${specifier} under ${instrumentationDefinitionCopy.moduleName}`
       )
+    } else if (format === 'module') {
+      registeredSpecifiers.set(url, specifier)
+      const modifiedUrl = new URL(url)
+      // add a query param to the resolved url so the load hook below knows
+      // to rewrite and wrap the source code
+      modifiedUrl.searchParams.set('hasNrInstrumentation', 'true')
+      resolvedModule.url = modifiedUrl.href
     } else {
-      logger.debug(`${specifier} is not a CommonJS module, skipping for now`)
+      logger.debug(`${specifier} is not a CommonJS nor ESM package, skipping for now.`)
     }
   }
 
   return resolvedModule
 }
 
+/**
+ * Hook chain responsible for determining how a URL should be interpreted, retrieved, and parsed.
+ *
+ * Our loader has to be the last user-supplied loader if chaining is happening,
+ * as we rely on `nextLoad` being the default Node.js resolve hook to load the ESM.
+ *
+ * Docs: https://nodejs.org/dist/latest-v18.x/docs/api/esm.html#loadurl-context-nextload
+ *
+ * @param {string} url the URL returned by the resolve chain
+ * @param {object} context metadata about the url, including conditions, format and import assertions
+ * @param {Function} nextLoad the subsequent load hook in the chain, or the Node.js default load hook after the last user-supplied load hook
+ * @returns {Promise} Promise object representing the resolution of a given specifier
+ */
+export async function load(url, context, nextLoad) {
+  if (!newrelic.agent || !isSupportedVersion()) {
+    return nextLoad(url, context, nextLoad)
+  }
+
+  let parsedUrl
+
+  try {
+    parsedUrl = new URL(url)
+  } catch (err) {
+    logger.error('Unable to parse url: %s, msg: %s', url, err.message)
+    return nextLoad(url, context, nextLoad)
+  }
+
+  const hasNrInstrumentation = parsedUrl.searchParams.get('hasNrInstrumentation')
+
+  if (!hasNrInstrumentation) {
+    return nextLoad(url, context, nextLoad)
+  }
+
+  /**
+   * undo the work we did in the resolve hook above
+   * so we can properly rewrite source and not get in an infinite loop
+   */
+  parsedUrl.searchParams.delete('hasNrInstrumentation')
+
+  const originalUrl = parsedUrl.href
+  const specifier = registeredSpecifiers.get(originalUrl)
+  const rewrittenSource = await wrapEsmSource(originalUrl, specifier)
+  logger.debug(`Registered module instrumentation for ${specifier}.`)
+
+  return {
+    format: 'module',
+    source: rewrittenSource,
+    shortCircuit: true
```

Comment:
Since we need to be the last hook in the chain, is the `shortCircuit` necessary? In my tests with `resolve`, it was not, but curious if `load` is different

Reply:
i dont' recall if this one was needed. i was fighting with so many issues that this may have been an attempt to fix and realized it made no effect, I'll test it out